src/dowel/csv_output.py

```diff
@@ -48,7 +50,35 @@ def record(self, data, prefix=''):
                            'Did you change key sets after your first '
                            'logger.log(TabularInput)?'.format(
                                set(self._fieldnames), set(to_csv.keys())))
-
+                
+                # close any existing log file
+                super().close()
+
+                # Read data from the existing log file once
+                with open(self._filename, 'r') as existing_file:
+                    reader = csv.DictReader(existing_file)
+                    file_data = []
+                    for row in reader:
+                        file_data.append(row)
+
+                    # Update header with new keys by union 
+                    self._joint_filednames = set(self._fieldnames)| set(to_csv.keys())
```

Comment:
I think you should actually just update the existing `_fieldnames` attribute, otherwise this re-write logic will be triggered every time there's the same "new" fields.

Reply:
My previous design is to create a new union field name `_joint_fieldname` and preserve the existing `_fieldnames` that may be useful to debug. 

But since you point out they may be redundant, I'll simply update the existing `_fieldnames` attribute.